include/swift/AST/DiagnosticsSema.def

```diff
@@ -394,6 +394,51 @@ ERROR(cannot_convert_closure_result_protocol,none,
 ERROR(cannot_convert_closure_result_nil,none,
       "'nil' is not compatible with closure result type %0", (Type))
 
+ERROR(cannot_convert_generic_type_argument,none,
+      "cannot convert %0 to expected argument type %1, "
+      "arguments to generic parameter %2 (%3 and %4) are expected to be equal",
+      (Type, Type, Identifier, Type, Type))
+ERROR(cannot_convert_generic_type_default_argument,none,
+      "default argument value of type %0 cannot be converted to type %1, "
+      "arguments to generic parameter %2 (%3 and %4) are expected to be equal",
+      (Type, Type, Identifier, Type, Type))
+ERROR(cannot_convert_generic_type_return,none,
+      "cannot convert return expression of type %0 to return type %1, "
+      "arguments to generic parameter %2 (%3 and %4) are expected to be equal",
+      (Type, Type, Identifier, Type, Type))
+ERROR(cannot_convert_generic_type_yield,none,
+      "cannot convert value of type %0 to expected yield type %1, "
+      "arguments to generic parameter %2 (%3 and %4) are expected to be equal",
+      (Type, Type, Identifier, Type, Type))
+ERROR(cannot_convert_generic_type_closure_result,none,
+      "cannot convert value of type %0 to closure result type %1, "
+      "arguments to generic parameter %2 (%3 and %4) are expected to be equal",
+      (Type, Type, Identifier, Type, Type))
+ERROR(cannot_convert_generic_type_array_element,none,
+      "cannot convert value of type %0 to expected element type %1, "
+      "arguments to generic parameter %2 (%3 and %4) are expected to be equal",
+      (Type, Type, Identifier, Type, Type))
+ERROR(cannot_convert_generic_type_dict_key,none,
+      "cannot convert value of type %0 to expected dictionary key type %1, "
```

Comment:
Some of these diagnostics could be combined using the select syntax.

Reply:
Yeah, that's why I decided against utilizing it here. It would make the types of the diagnostic not line up so we wouldn't be able to unify the logic for actually producing the error message. As such, I think that's a trade off that's not worth making. I do wish I could make use of the select syntax here though but I think the benefits of having a single logical flow here outweigh the benefits of having less diagnostics relating to this case.